### PR TITLE
Dimer Maker

### PR DESCRIPTION
**Pull request by ivanorman**
Adds dimer maker module and associated unit tests. Currently the dimer module is in src/ghostfragment/fragmenting as I wasn't sure where to put it. I wasn't sure how the case for a single fragment should be handled, but currently I have it return an empty FragmentedNuclei.

### CODE REVIEW

File: tests/cxx/unit_tests/ghostfragment/fragmenting/dimer_maker.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    auto mm        = initialize();
+    auto& mod      = mm.at("Dimer Maker");
+
+    SECTION("Empty Nuclei") {
+        input_type nukes;
+        result_type corr(nukes);
+        mod.change_submod("Fragmentation", make_frag_module(nukes, corr));
+        result_type frags = mod.run_as<my_pt>(nukes);
+        REQUIRE(corr.operator==(frags));
```
Comment (ryanmrichard) [inline tests/cxx/unit_tests/ghostfragment/fragmenting/dimer_maker.cpp:34]:
```suggestion
        REQUIRE(corr == frags);
```
You shouldn't have to call the member function explicitly like that.

File: tests/cxx/unit_tests/ghostfragment/fragmenting/dimer_maker.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        chemist::Molecule mol;
+        mol.push_back(chemist::Atom("H", 1, 1837.289, 0, 0, 0));
+        input_type nukes(mol.nuclei());
+        result_type fragments(nukes);
+        result_type corr(nukes);
+        fragments.add_fragment({0});
+
+        mod.change_submod("Fragmentation", make_frag_module(nukes, fragments));
+        result_type frags = mod.run_as<my_pt>(nukes);
+        REQUIRE(corr.operator==(frags));
```
Comment (ryanmrichard) [inline tests/cxx/unit_tests/ghostfragment/fragmenting/dimer_maker.cpp:47]:
```suggestion
        REQUIRE(corr == frags);
```

**Review by ryanmrichard — CHANGES_REQUESTED** [review]
I forgot to mention, that like our other fragmenters, we don't want this module to return dimers which are subsets of other dimers. So I would recommend adding another test case:

```.cpp
    SECTION("Methane (alternative fragmentation)") {
        input_type nukes(hydrocarbon(1).nuclei());
        result_type fragments(nukes);
        fragments.add_fragment({0, 1});
        fragments.add_fragment({1, 3});
        fragments.add_fragment({2, 3});
        fragments.add_fragment({0, 4});

        result_type corr(nukes);
        corr.add_fragment({0, 1, 2, 3});
        corr.add_fragment({0, 1, 3, 4});
        corr.add_fragment({0, 2, 3, 4});

        mod.change_submod("Fragmentation", make_frag_module(nukes, fragments));
        result_type frags = mod.run_as<my_pt>(nukes);
        REQUIRE(corr.operator==(frags));
    }
```

**Comment by ryanmrichard** [comment]
Also, in the future make sure you start PRs off of an up-to-date version of the master branch; you presumably started this PR off of an out-of-date version, which is why the commit history is so long.

**Comment by ivanorman** [comment]
Fixed so doesn't return fragments that are proper subsets of other fragments, added relevant testing.

**Review by ryanmrichard — APPROVED** [review]

**Pull request merged**